File: backend/src/routes/pracownicy.js
'use strict';
const express = require('express');
const { body } = require('express-validator');
const db = require('../config/db');
const { authenticate } = require('../middleware/auth');
const { roleGuard } = require('../middleware/roleGuard');
const { validate } = require('../middleware/validate');
const { writeAudit } = require('../services/auditService');

const router = express.Router();

router.get('/', authenticate, async (req, res) => {
  const { tenant_id, komorka_id, aktywny, q } = req.query;
  const { rola, tenantId } = req.user;
  let query = `SELECT p.id, p.uzytkownik_id, p.tenant_id, p.komorka_id,
                      p.stanowisko, p.data_zatrudnienia, p.data_zwolnienia, p.aktywny,
                      p.przelozony_id,
                      u.imie, u.nazwisko, u.email, u.username, u.rola,
                      k.nazwa as komorka_nazwa, t.nazwa as tenant_nazwa,
                      pr.id as przel_id,
                      pu.imie || ' ' || pu.nazwisko as przel_nazwa
                 FROM pracownicy p
                 JOIN uzytkownicy u ON u.id = p.uzytkownik_id
                 LEFT JOIN komorki_org k ON k.id = p.komorka_id
                 LEFT JOIN tenants t ON t.id = p.tenant_id
                 LEFT JOIN pracownicy pr ON pr.id = p.przelozony_id
                 LEFT JOIN uzytkownicy pu ON pu.id = pr.uzytkownik_id
                WHERE 1=1`;
  const params = [];

  // Tenant scope
  const effectiveTenantId = tenant_id || (['KIEROWNIK','PRACOWNIK'].includes(rola) ? tenantId : null);
  if (effectiveTenantId) { params.push(effectiveTenantId); query += ` AND p.tenant_id = $${params.length}`; }
  if (komorka_id) { params.push(komorka_id); query += ` AND p.komorka_id = $${params.length}`; }
  if (aktywny !== undefined) { params.push(aktywny === 'true'); query += ` AND p.aktywny = $${params.length}`; }
  if (q) { params.push(`%${q}%`); query += ` AND (u.imie ILIKE $${params.length} OR u.nazwisko ILIKE $${params.length} OR u.email ILIKE $${params.length})`; }

  query += ' ORDER BY u.nazwisko, u.imie';
  const { rows } = await db.query(query, params);
  res.json(rows);
});

router.get('/:id', authenticate, async (req, res) => {
  const { rows } = await db.query(
    `SELECT p.*, u.imie, u.nazwisko, u.email, u.username, u.rola,
            k.nazwa as komorka_nazwa, t.nazwa as tenant_nazwa,
            pu.imie || ' ' || pu.nazwisko as przel_nazwa,
            pr.stanowisko as przel_stanowisko
       FROM pracownicy p
       JOIN uzytkownicy u ON u.id = p.uzytkownik_id
       LEFT JOIN komorki_org k ON k.id = p.komorka_id
       LEFT JOIN tenants t ON t.id = p.tenant_id
       LEFT JOIN pracownicy pr ON pr.id = p.przelozony_id
       LEFT JOIN uzytkownicy pu ON pu.id = pr.uzytkownik_id
      WHERE p.id = $1`,
    [req.params.id]
  );
  if (!rows.length) return res.status(404).json({ error: 'Pracownik nie istnieje.' });

  // Get active permissions count
  const { rows: permRows } = await db.query(
    'SELECT COUNT(*) as cnt FROM uprawnienia WHERE pracownik_id=$1 AND aktywne=TRUE',
    [req.params.id]
  );
  res.json({ ...rows[0], liczba_uprawnien: parseInt(permRows[0].cnt) });
});

router.post('/', authenticate, roleGuard('IT_ADMIN', 'SUPERADMIN', 'KADRY'),
  [
    body('uzytkownik_id').isInt(),
    body('tenant_id').isInt(),
  ],
  validate,
  async (req, res) => {
    const { uzytkownik_id, tenant_id, komorka_id, stanowisko, data_zatrudnienia, przelozony_id } = req.body;
    const { rows } = await db.query(
      `INSERT INTO pracownicy (uzytkownik_id, tenant_id, komorka_id, stanowisko, data_zatrudnienia, przelozony_id)
       VALUES ($1,$2,$3,$4,$5,$6) RETURNING *`,
      [uzytkownik_id, tenant_id, komorka_id || null, stanowisko || null, data_zatrudnienia || null, przelozony_id || null]
    );
    await writeAudit({ userId: req.user.userId, username: req.user.username, rola: req.user.rola, akcja: 'PRACOWNIK_CREATE', tabelaDocelowa: 'pracownicy', rekordId: rows[0].id, noweDane: req.body });
    res.status(201).json(rows[0]);
  }
);

router.put('/:id', authenticate, roleGuard('IT_ADMIN', 'SUPERADMIN', 'KADRY'), async (req, res) => {
  const { komorka_id, stanowisko, data_zatrudnienia, data_zwolnienia, aktywny, przelozony_id } = req.body;
  const { rows: old } = await db.query('SELECT * FROM pracownicy WHERE id=$1', [req.params.id]);
  if (!old.length) return res.status(404).json({ error: 'Pracownik nie istnieje.' });

  const { rows } = await db.query(
    `UPDATE pracownicy SET komorka_id=$1, stanowisko=$2, data_zatrudnienia=$3,
       data_zwolnienia=$4, aktywny=$5, przelozony_id=$6 WHERE id=$7 RETURNING *`,
    [komorka_id || null, stanowisko, data_zatrudnienia, data_zwolnienia || null, aktywny !== false, przelozony_id || null, req.params.id]
  );

  // If termination date set and newly set → initiate departure review
  if (data_zwolnienia && !old[0].data_zwolnienia) {
    // This would trigger an ODEJSCIE review in a production system
    await writeAudit({ userId: req.user.userId, username: req.user.username, rola: req.user.rola, akcja: 'PRACOWNIK_ODEJSCIE', tabelaDocelowa: 'pracownicy', rekordId: req.params.id, noweDane: { data_zwolnienia } });
  }

  await writeAudit({ userId: req.user.userId, username: req.user.username, rola: req.user.rola, akcja: 'PRACOWNIK_UPDATE', tabelaDocelowa: 'pracownicy', rekordId: req.params.id, noweDane: req.body, stareDane: old[0] });
  res.json(rows[0]);
});

// Get subordinates
router.get('/:id/podwladni', authenticate, async (req, res) => {
  const { rows } = await db.query(
    `SELECT p.id, u.imie, u.nazwisko, u.email, p.stanowisko, k.nazwa as komorka_nazwa,
            p.aktywny,
            (SELECT COUNT(*) FROM uprawnienia up WHERE up.pracownik_id = p.id AND up.aktywne = TRUE) as liczba_uprawnien
       FROM pracownicy p
       JOIN uzytkownicy u ON u.id = p.uzytkownik_id
       LEFT JOIN komorki_org k ON k.id = p.komorka_id
      WHERE p.przelozony_id = $1 AND p.aktywny = TRUE
      ORDER BY u.nazwisko`,
    [req.params.id]
  );
  res.json(rows);
});

module.exports = router;

File: backend-py/routes/pracownicy.py
from __future__ import annotations
from typing import Optional

from fastapi import APIRouter, Depends, HTTPException, Request
from schemas import BaseSchema
from sqlalchemy import text
from sqlalchemy.orm import Session

from database import get_session
from dependencies import CurrentUser, get_current_user, require_roles, write_audit, get_client_ip

router = APIRouter()

# Columns used to build a display name — falls back to p.imie/nazwisko when no account
_NAME_COLS = """
    COALESCE(u.imie,     p.imie)     AS imie,
    COALESCE(u.nazwisko, p.nazwisko) AS nazwisko,
    COALESCE(u.email,    p.email)    AS email,
    u.username, u.rola
"""


class PracownikIn(BaseSchema):
    uzytkownik_id: Optional[int] = None   # null => pracownik bez konta systemowego
    tenant_id: int
    jednostka_id: Optional[int] = None
    stanowisko: Optional[str] = None
    # used when uzytkownik_id is None
    imie: Optional[str] = None
    nazwisko: Optional[str] = None
    email: Optional[str] = None
    data_zatrudnienia: Optional[str] = None
    data_zwolnienia: Optional[str] = None
    aktywny: Optional[bool] = True
    przelozony_id: Optional[int] = None


@router.get("")
async def list_pracownicy(tenant_id: Optional[int] = None, jednostka_id: Optional[int] = None,
                           aktywny: Optional[str] = None, q: Optional[str] = None,
                           session: Session = Depends(get_session),
                           current_user: CurrentUser = Depends(get_current_user)):
    query = f"""SELECT p.id, p.uzytkownik_id, p.tenant_id, p.jednostka_id,
                      p.stanowisko, p.data_zatrudnienia, p.data_zwolnienia, p.aktywny,
                      p.przelozony_id,
                      {_NAME_COLS},
                      j.nazwa as jednostka_nazwa, t.nazwa as tenant_nazwa,
                      pr.id as przel_id,
                      COALESCE(pu.imie, prd.imie) || ' ' || COALESCE(pu.nazwisko, prd.nazwisko) as przel_nazwa
                 FROM pracownicy p
                 LEFT JOIN uzytkownicy u   ON u.id  = p.uzytkownik_id
                 LEFT JOIN jednostki_org j ON j.id  = p.jednostka_id
                 LEFT JOIN tenants t        ON t.id  = p.tenant_id
                 LEFT JOIN pracownicy pr    ON pr.id = p.przelozony_id
                 LEFT JOIN uzytkownicy pu   ON pu.id = pr.uzytkownik_id
                 LEFT JOIN pracownicy prd   ON prd.id = pr.id
                WHERE 1=1"""
    params: dict = {}

    effective_tenant = tenant_id
    if current_user.rola in ("KIEROWNIK", "PRACOWNIK"):
        effective_tenant = effective_tenant or current_user.tenantId

    if effective_tenant:
        query += " AND p.tenant_id=:tid"
        params["tid"] = effective_tenant
    if jednostka_id:
        query += " AND p.jednostka_id=:jid"
        params["jid"] = jednostka_id
    if aktywny is not None:
        query += " AND p.aktywny=:a"
        params["a"] = aktywny == "true"
    if q:
        query += """ AND (COALESCE(u.imie,p.imie) LIKE :q
                       OR COALESCE(u.nazwisko,p.nazwisko) LIKE :q
                       OR COALESCE(u.email,p.email) LIKE :q)"""
        params["q"] = f"%{q}%"

    query += " ORDER BY COALESCE(u.nazwisko,p.nazwisko), COALESCE(u.imie,p.imie)"
    rows = session.execute(text(query), params).mappings().all()
    return [dict(r) for r in rows]


@router.get("/:id/podwladni", include_in_schema=False)
@router.get("/{prac_id}/podwladni")
async def get_podwladni(prac_id: int, session: Session = Depends(get_session),
                         current_user: CurrentUser = Depends(get_current_user)):
    rows = session.execute(
        text(f"""SELECT p.id, {_NAME_COLS}, p.stanowisko, j.nazwa as jednostka_nazwa,
                       p.aktywny,
                       (SELECT COUNT(*) FROM uprawnienia up WHERE up.pracownik_id=p.id AND up.aktywne=1) as liczba_uprawnien
                  FROM pracownicy p
                  LEFT JOIN uzytkownicy u   ON u.id = p.uzytkownik_id
                  LEFT JOIN jednostki_org j ON j.id = p.jednostka_id
                 WHERE p.przelozony_id=:id AND p.aktywny=1
                 ORDER BY COALESCE(u.nazwisko, p.nazwisko)"""),
        {"id": prac_id},
    ).mappings().all()
    return [dict(r) for r in rows]


@router.get("/{prac_id}")
async def get_pracownik(prac_id: int, session: Session = Depends(get_session),
                         current_user: CurrentUser = Depends(get_current_user)):
    row = session.execute(
        text(f"""SELECT p.*, {_NAME_COLS},
                       j.nazwa as jednostka_nazwa, t.nazwa as tenant_nazwa,
                       COALESCE(pu.imie,prd.imie) || ' ' || COALESCE(pu.nazwisko,prd.nazwisko) as przel_nazwa,
                       pr.stanowisko as przel_stanowisko
                  FROM pracownicy p
                  LEFT JOIN uzytkownicy u   ON u.id  = p.uzytkownik_id
                  LEFT JOIN jednostki_org j ON j.id  = p.jednostka_id
                  LEFT JOIN tenants t        ON t.id  = p.tenant_id
                  LEFT JOIN pracownicy pr    ON pr.id = p.przelozony_id
                  LEFT JOIN uzytkownicy pu   ON pu.id = pr.uzytkownik_id
                  LEFT JOIN pracownicy prd   ON prd.id = pr.id
                 WHERE p.id=:id"""),
        {"id": prac_id},
    ).mappings().first()
    if not row:
        raise HTTPException(status_code=404, detail="Pracownik nie istnieje.")
    result = dict(row)
    cnt = session.execute(
        text("SELECT COUNT(*) as cnt FROM uprawnienia WHERE pracownik_id=:id AND aktywne=1"),
        {"id": prac_id},
    ).mappings().first()
    result["liczba_uprawnien"] = cnt["cnt"] if cnt else 0
    return result


@router.post("", status_code=201)
async def create_pracownik(body: PracownikIn, request: Request,
                            session: Session = Depends(get_session),
                            current_user: CurrentUser = Depends(require_roles("IT_ADMIN", "SUPERADMIN", "KADRY"))):
    if not body.uzytkownik_id and not body.imie:
        raise HTTPException(status_code=422, detail="Podaj imię i nazwisko lub powiąż konto użytkownika.")
    row = session.execute(
        text("""INSERT INTO pracownicy
                    (uzytkownik_id, tenant_id, jednostka_id, stanowisko,
                     imie, nazwisko, email,
                     data_zatrudnienia, przelozony_id, aktywny)
                VALUES (:uid,:tid,:jid,:s,:imie,:nazwisko,:email,:d,:p,1) RETURNING *"""),
        {"uid": body.uzytkownik_id, "tid": body.tenant_id, "jid": body.jednostka_id,
         "s": body.stanowisko,
         "imie": body.imie, "nazwisko": body.nazwisko, "email": body.email,
         "d": body.data_zatrudnienia, "p": body.przelozony_id},
    ).mappings().first()
    session.commit()
    result = dict(row)
    write_audit(session, user_id=current_user.userId, username=current_user.username,
                rola=current_user.rola, akcja="PRACOWNIK_CREATE", tabela_docelowa="pracownicy",
                rekord_id=result["id"], nowe_dane=body.model_dump(), ip_adres=get_client_ip(request))
    return result


@router.put("/{prac_id}")
async def update_pracownik(prac_id: int, body: PracownikIn, request: Request,
                            session: Session = Depends(get_session),
                            current_user: CurrentUser = Depends(require_roles("IT_ADMIN", "SUPERADMIN", "KADRY"))):
    old = session.execute(text("SELECT * FROM pracownicy WHERE id=:id"), {"id": prac_id}).mappings().first()
    if not old:
        raise HTTPException(status_code=404, detail="Pracownik nie istnieje.")
    old = dict(old)

    row = session.execute(
        text("""UPDATE pracownicy
                   SET jednostka_id=:jid, stanowisko=:s,
                       imie=:imie, nazwisko=:nazwisko, email=:email,
                       data_zatrudnienia=:dt, data_zwolnienia=:dz,
                       aktywny=:a, przelozony_id=:p
                 WHERE id=:id RETURNING *"""),
        {"jid": body.jednostka_id, "s": body.stanowisko,
         "imie": body.imie, "nazwisko": body.nazwisko, "email": body.email,
         "dt": body.data_zatrudnienia, "dz": body.data_zwolnienia,
         "a": body.aktywny if body.aktywny is not None else True,
         "p": body.przelozony_id, "id": prac_id},
    ).mappings().first()
    session.commit()
    result = dict(row)

    if body.data_zwolnienia and not old.get("data_zwolnienia"):
        write_audit(session, user_id=current_user.userId, username=current_user.username,
                    rola=current_user.rola, akcja="PRACOWNIK_ODEJSCIE", tabela_docelowa="pracownicy",
                    rekord_id=prac_id, nowe_dane={"data_zwolnienia": body.data_zwolnienia},
                    ip_adres=get_client_ip(request))

    write_audit(session, user_id=current_user.userId, username=current_user.username,
                rola=current_user.rola, akcja="PRACOWNIK_UPDATE", tabela_docelowa="pracownicy",
                rekord_id=prac_id, nowe_dane=body.model_dump(), stare_dane=old,
                ip_adres=get_client_ip(request))
    return result
